### PR TITLE
WIP:IEP-373 Application Level Tracing

### PR DESCRIPTION
**Pull request by sigmaaa**
Need to add: 
- error handling 
- automatically openocd start

### CODE REVIEW

**Comment by sigmaaa** [comment]
need to update documentation.

File: bundles/com.espressif.idf.debug.gdbjtag.openocd/plugin.xml [hunk 1]
```diff
@@ -125,4 +125,13 @@
           id="com.espressif.idf.debug.gdbjtag.openocd.heaptracing.HeapTracingBreakpointActionPage">
     </actionPage>
  </extension>
+ <extension
+       point="org.eclipse.ui.commands">
+      <command
+          defaultHandler="com.espressif.idf.debug.gdbjtag.openocd.ui.AppLvlTracingHandler"
+          id="com.espressif.idf.ui.appLvlTracing"
+          name="Application Level Tracing">
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.debug.gdbjtag.openocd/plugin.xml:133]: Externalise string

File: bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/ui/AppLvlTracingHandler.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+@SuppressWarnings("restriction")
+public class AppLvlTracingHandler extends AbstractHandler
+{
+
+	private IResource project;
+
+	@Override
+	public Object execute(ExecutionEvent event) throws ExecutionException
+	{
+		Shell activeShell = PlatformUI.getWorkbench().getActiveWorkbenchWindow().getShell();
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/ui/AppLvlTracingHandler.java:43]: Maybe we could use EclispeUtil.getShell()

File: bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/ui/AppLvlTracingHandler.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+		IResource projectOut = EclipseHandler.getSelectedProject(IPageLayout.ID_PROJECT_EXPLORER);
+		project = projectOut;
+		if (projectOut == null)
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/ui/AppLvlTracingHandler.java:46]: Can we use `project` itself instead one more variable `projectOut`

File: bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/ui/messages.properties [hunk 4]
```diff
@@ -392,3 +392,7 @@ IDFLaunchTargetNotFoundMsg3=Do you want to create a new IDF launch target?
 
 StartupTabOpenOcdGroup=OpenOCD Options
 StartupTabFlashBeforeStart=Flash every time with application binaries
+MissingDebugConfigurationTitle=Missing debug configuration
+OpenOcdFailedMsg=OpenOCD was not started
+DebugConfigurationNotFoundMsg=No matching debug configuration was found for the selected project. Do you want to create it?
+AppLvlTracingJob="Launching application level tracing"
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/ui/messages.properties:398]: Double quotes..is it intentional?

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.net.Socket;
+
+import com.espressif.idf.core.logging.Logger;
+
+public class TclClient {
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java:12]: Add author info

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
+package com.espressif.idf.ui.tracing;
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java:1]: copyright info missing

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+	public void startTracing(String[] params)
+	{
+		String startCommand = "esp apptrace start "; //$NON-NLS-1$
+		startCommand = startCommand + String.join(" ", params); //$NON-NLS-1$
+    	if(clientSocket.isConnected()) {
+    		try {
+				out.write("capture \"" + startCommand + "\""); //$NON-NLS-1$ //$NON-NLS-2$
+				out.write(0x1a);
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java:35]: Can you add a comment on why `0x1a` is required for future reference

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java:81]: maybe we could remove empty lines

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/messages.properties [hunk 9]
```diff
@@ -32,3 +32,24 @@ TracingCallerView_ColAddress=Address
...
+AppLvlTracing_StopTmo=Stop Timeout:
+AppLvlTracing_BytesToSKip=Bytes to Skip:
+AppLvlTracing_OutFile=Out File:
+AppLvlTracing_TraceScript=Trace Processing Script:
+AppLvlTracing_StartParse=Start parse
+AppLvlTracing_ConsoleName=App Tracing Console
+AppLvlTracing_ScriptBrowseLbl=Select script file
+AppLvlTracing_StartParsingCommandLbl=Start Parsing Command:
+AppLvlTracing_DumpFileNotExistsInfo=The trace log file was not found. Provide parameters to the esp start tracing command to start application-level tracing. This will create a trace log file
+AppLvlTracing_NotValidDirectoryPath=  directory path is not valid
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/messages.properties:53]: Empty space before message?

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/messages.properties [hunk 10]
```diff
@@ -32,3 +32,24 @@ TracingCallerView_ColAddress=Address
...
+AppLvlTracing_NotValidScriptFilePath=  file was not found
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/messages.properties:54]: Empty space before the message?

File: bundles/com.espressif.idf.core/src/com/espressif/idf/core/build/CommandEntry.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
+package com.espressif.idf.core.build;
+
+/**
+ * Represents a parsed command entry of a compile_commands.json file.
+ * 
+ * @author weber
+ */
+class CommandEntry
```
Comment (alirana01) [inline bundles/com.espressif.idf.core/src/com/espressif/idf/core/build/CommandEntry.java:8]: Do we need this file? 🤔

File: bundles/com.espressif.idf.core/src/com/espressif/idf/core/build/IDFBuildConfiguration.java [hunk 12]
```diff
@@ -88,32 +96,32 @@
...
+	 * Cmake-generated build scripts re-run cmake if one of the CMakeLists.txt files was modified, but that output goes
+	 * through ErrorParserManager and is impossible to parse because cmake outputs to both stderr and stdout and
+	 * ErrorParserManager intermixes these streams making it impossible to parse for errors.<br>
+	 * To work around that, we run cmake in advance with its dedicated working error parser.
 	 */
 	private boolean cmakeListsModified;
 	private ICMakeToolChainFile toolChainFile;
 
-	public IDFBuildConfiguration(IBuildConfiguration config, String name) throws CoreException {
+	public IDFBuildConfiguration(IBuildConfiguration config, String name) throws CoreException
```
Comment (alirana01) [inline bundles/com.espressif.idf.core/src/com/espressif/idf/core/build/IDFBuildConfiguration.java:109]: Is the master rebase properly done here?

**Review by alirana01 — CHANGES_REQUESTED** [review]
Is the branch properly rebased or merged?

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java (hunk 6 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java:5]: added

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java (hunk 5 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java:21]: done

File: bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/ui/AppLvlTracingHandler.java (hunk 2 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/ui/AppLvlTracingHandler.java]: done

File: bundles/com.espressif.idf.debug.gdbjtag.openocd/plugin.xml (hunk 1 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.debug.gdbjtag.openocd/plugin.xml]: fixed

File: bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/ui/AppLvlTracingHandler.java (hunk 3 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/ui/AppLvlTracingHandler.java]: fixed

File: bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/ui/messages.properties (hunk 4 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/ui/messages.properties]: double quotes removed

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java (hunk 5 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java:21]: added

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java (hunk 6 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java:5]: added

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java (hunk 7 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java]: added a variable for EOL character - openOcdEOL and comment to it

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java (hunk 8 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/TclClient.java]: removed

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/messages.properties (hunk 9 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/messages.properties]: fixed

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/messages.properties (hunk 10 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/messages.properties]: fixed

**Comment by sigmaaa** [comment]
> Is the branch properly rebased or merged?

after rebase I accidentally synchronized the branch, this caused a problem. Fixed

**Review by kolipakakondal — APPROVED** [review]
LGTM. Good job @sigmaaa

**Pull request merged**